.teamcity/src/main/kotlin/model/FunctionalTestBucketGenerator.kt

```diff
@@ -0,0 +1,213 @@
+
+    fun generate(jsonFile: File) {
+        jsonFile.writeText(JSON.toJSONString(buckets.map {
+            TestCoverageAndBucketSplits(it.key.uuid, it.value.map { it.toJsonBucket() })
+        }, SerializerFeature.PrettyFormat))
+    }
+
+    private
+    fun buildBuckets(buildClassTimeJson: File, model: CIBuildModel): Map<TestCoverage, List<BuildTypeBucket>> {
+        val jsonObj = JSON.parseObject(buildClassTimeJson.readText()) as JSONObject
+        val buildProjectClassTimes: BuildProjectToSubprojectTestClassTimes = jsonObj.map { buildProjectToSubprojectTestClassTime ->
+            buildProjectToSubprojectTestClassTime.key to (buildProjectToSubprojectTestClassTime.value as JSONObject).map { subProjectToTestClassTime ->
+                subProjectToTestClassTime.key to (subProjectToTestClassTime.value as JSONArray).map { TestClassTime(it as JSONObject) }
+            }.toMap()
+        }.toMap()
+
+        val result = mutableMapOf<TestCoverage, List<BuildTypeBucket>>()
+        for (stage in model.stages) {
+            for (testCoverage in stage.functionalTests) {
+                if (testCoverage.testType !in listOf(TestType.allVersionsCrossVersion, TestType.quickFeedbackCrossVersion)) {
+                    result[testCoverage] = splitBucketsByTestClassesForBuildProject(testCoverage, stage, buildProjectClassTimes)
+                }
+            }
+        }
+        return result
+    }
+
+    private
+    fun splitBucketsByTestClassesForBuildProject(testCoverage: TestCoverage, stage: Stage, buildProjectClassTimes: BuildProjectToSubprojectTestClassTimes): List<BuildTypeBucket> {
+        val validSubprojects = model.subprojects.getSubprojectsFor(testCoverage, stage)
+
+        // Build project not found, don't split into buckets
+        val subProjectToClassTimes: MutableMap<String, List<TestClassTime>> = determineSubProjectClassTimes(testCoverage, buildProjectClassTimes)?.toMutableMap() ?: return validSubprojects
+
+        validSubprojects.forEach {
+            if (!subProjectToClassTimes.containsKey(it.name)) {
+                subProjectToClassTimes[it.name] = emptyList()
+            }
+        }
+
+        val subProjectTestClassTimes: List<SubprojectTestClassTime> = subProjectToClassTimes
+            .entries
+            .filter { "UNKNOWN" != it.key }
+            .filter { model.subprojects.getSubprojectByName(it.key) != null }
+            .map { SubprojectTestClassTime(model.subprojects.getSubprojectByName(it.key)!!, it.value.filter { it.sourceSet != "test" }) }
+            .sortedBy { -it.totalTime }
+
+        return when {
+            testCoverage.testType == TestType.platform && testCoverage.os == Os.WINDOWS -> {
+                splitDocsSubproject(validSubprojects) + splitIntoBucketsExcludingSpecialBuckets(listOf("docs"), validSubprojects, subProjectTestClassTimes, testCoverage)
+            }
+            testCoverage.os == Os.LINUX -> {
+                splitIntoBuckets(
```

Comment:
Does this mean we are running `:docs:docsTest` now via test distribution?

Reply:
I moved `docs:docsTest` to several special buckets now.